velox/functions/sparksql/DecimalVectorFunctions.cpp

```diff
@@ -0,0 +1,345 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/expression/DecodedArgs.h"
+#include "velox/expression/VectorFunction.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+// Rescale two inputs as the same scale and compare. Returns 0 when a is equal
+// with b. Returns -1 when a is less than b. Returns 1 when a is greater than b.
+template <typename T>
+int32_t rescaleAndCompare(T a, T b, int32_t deltaScale) {
```

Comment:
This is a vector function, which receives a batch of rows at a time. All values have same precision and scale, hence, we do not need to re-scale values one at a time. Instead, we can (1) rescale all 'a' values; (2) rescale all 'b' values; (3) compare rescaled values. This may be more efficient than doing this one value at a time. It would be nice to add a benchmark to help decide on best implementation.

Reply:
@majetideepak @mbasmanova Some cases cannot be handled with cast. For example, when the two inputs are a (DECIMAL(38, 34)) and b (DECIMAL(6, 1)), if we cast b to scale 34, its precision will become 39 and cannot fit in a decimal. If we cast a to scale 1, precision loss happens and could affect the result.

Reply:
This issue can happen with the rescaling here also correct?
 If we take 
`a (DECIMAL(38, 37)) = 1.99..(total 37)`  and `b (DECIMAL(38, 0)) = 9999..(total38)`.
Now deltaScale = 37.
 `bScaled = b * velox::DecimalUtil::kPowersOfTen[deltaScale];` will overflow?

Reply:
If we rescale `b (DECIMAL(38, 0)) = 9999..(total38)` to `DECIMAL(38, 37)`, the new precision will become 75. In this case, int256 type is needed. Because the maximum of int256 is `2^256/2-1`, which is of scale 77 in decimal, so overflow will not happen.
The maximun of int256 in decimal: `57896044618658097711785492504343953926634992332820282019728792003956564819967`.

Reply:
> It would be nice to add a benchmark to help decide on best implementation.

@mbasmanova I added CompareBenchmark in this PR, and compared below two implementations:
1) Rescale during comparison (the implementation of this PR).
2) Rescale with cast first, and then compare.

Below is the result, and we can see 1 is better than 2 in performance. Also in view of the limitation of 2 as discussed in https://github.com/facebookincubator/velox/pull/6207#discussion_r1308779033, can we keep this PR?
```
============================================================================
[...]rksql/benchmarks/CompareBenchmark.cpp     relative  time/iter   iters/s
============================================================================
Gt                                                         20.97ns    47.69M
Gt_with_cast                                               27.18ns    36.80M
Gte                                                        21.16ns    47.27M
Gte_with_cast                                              27.79ns    35.98M
Lt                                                         21.04ns    47.52M
Lt_with_cast                                               27.09ns    36.91M
Lte                                                        20.89ns    47.87M
Lte_with_cast                                              27.34ns    36.58M
Eq                                                         20.30ns    49.27M
Eq_with_cast                                               26.97ns    37.08M
Neq                                                        20.95ns    47.73M
```